CMakeLists.txt

```diff
@@ -540,30 +495,52 @@ IF (BUILD_AFQMC AND NOT APPLE)
   LINK_LIBRARIES("rt")
 ENDIF()
 
-# setup ctest variables
-IF ( HAVE_MPI )
-  SET( USE_MPI 1 )
-  IF ( NOT DEFINED MPIEXEC )
-    SET( MPIEXEC "mpirun" )
-  ENDIF()
-
-  IF ( NOT DEFINED MPIEXEC_NUMPROC_FLAG )
-    SET( MPIEXEC_NUMPROC_FLAG "-np" )
+#-------------------------------------------------------------------
+#  check MPI installation
+#-------------------------------------------------------------------
+IF(QMC_MPI)
+  # for backward compatibility with MPIEXEC
+  IF(DEFINED MPIEXEC AND NOT DEFINED MPIEXEC_EXECUTABLE)
+    MESSAGE(WARNING "Setting MPIEXEC_EXECUTABLE based on MPIEXEC. MPIEXEC has been deprecated! Use MPIEXEC_EXECUTABLE instead!")
+    SET(MPIEXEC_EXECUTABLE ${MPIEXEC})
+  ENDIF(DEFINED MPIEXEC AND NOT DEFINED MPIEXEC_EXECUTABLE)
+  SET(MPI_CXX_SKIP_MPICXX TRUE)
+  SET(MPI_DETERMINE_LIBRARY_VERSION TRUE)
+  FIND_PACKAGE(MPI COMPONENTS CXX)
+
+  IF(NOT MPI_FOUND)
+    MESSAGE(FATAL_ERROR "MPI support not found! Provide MPI compiler wrappers or build without MPI by passing '-DQMC_MPI=0' to cmake.")
+  ENDIF(NOT MPI_FOUND)
+
+  IF(${MPI_CXX_LIBRARY_VERSION_STRING} MATCHES "MVAPICH2" AND NOT ${MPI_CXX_LIBRARY_VERSION_STRING} MATCHES "disable-registration-cache")
+    MESSAGE(FATAL_ERROR "MVAPICH2 with registration cache enabled breaks QMCPACK. "
+                        "Use a different MPI library or build MVAPICH2 with --disable-registration-cache configure option.")
   ENDIF()
 
-  IF ( NOT DEFINED MPIEXEC_PREFLAGS )
-    SET( MPIEXEC_PREFLAGS "" )
+  IF(NOT ${CMAKE_CXX_COMPILER} STREQUAL ${MPI_CXX_COMPILER})
+    SET(MPI_WARNING_LIST
+        "Building MPI version without using MPI compiler wrappers.\n"
```

Comment:
I think the preferred way to use CMake with MPI is to set the the compiler variables to the underlying compiler, and let FindMPI find all the include paths and libraries, etc.   Is there something in particular that we know will break QMCPACK (like not all the include or libraries paths are specified in a way that will pick up the MPI files?).     I don't think anything should be done on this PR, just pointing it out here.

Reply:
The safe way to use MPI with cmake is to use the MPI compiler wrappers. This give a guaranteed outcome and the correct libraries will be linked. We recommend this route in the manual. Mileage may vary the other way, particularly on a "legacy" machine with multiple dusty MPI installations.

Reply:
IMO, the preferred way is passing MPI compiler wrappers as compilers. FindMPI knows how to deal with this and keep the compile/link line clean. If the compilers passed in are not compiler wrappers, FindMPI goes to the folder containing MPIEXEC_EXECUTABLE and look for compiler wrappers of all the needed languages. Although it probably can find all the libraries and set proper flags and include directories based on querying MPIEXEC_EXECUTABLE options like -echo/-showme, the resulting compile and link lines get extremely long and error prone.